internal/measurexlite/resolver.go

```diff
@@ -0,0 +1,106 @@
+		URL:                               URL,
+		NewParallelDNSOverHTTPSResolverFn: nil,
+		ResolverSystem:                    netxlite.NewStdlibResolver(logger),
+	}
+}
+
+// TrustedRecursiveResolver2 emulates Firefox's TRR2 mode
+type TrustedRecursiveResolver2 struct {
+	Logger model.Logger
+	URL    string
+
+	// NewParallelDNSOverHTTPSResolverFn is OPTIONAL and can be used to overide
+	// calls to the netxlite.NewParallelDNSOverHTTPSResolver factory
+	NewParallelDNSOverHTTPSResolverFn func(model.Logger, string) model.Resolver
+
+	// ResolverSystem is MANADATORY and is used as a fallback if the DoH resolver fails
+	ResolverSystem model.Resolver
+}
+
+var _ model.Resolver = &TrustedRecursiveResolver2{}
+
+// Address implements model.Resolver.Address
+// we can't return a decisive address here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Address() string {
+	return ""
+}
+
+// Network implements model.Resolver.Network
+// we can't return a decisive network here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Network() string {
+	return ""
+}
+
+// CloseIdleConnections implements model.Resolver.CloseIdleConnections
+// we only check the fallback resolver since DoH resolver is generated on the fly
+func (trr TrustedRecursiveResolver2) CloseIdleConnections() {
+	trr.ResolverSystem.CloseIdleConnections()
+}
+
+// LookupHost implements model.Resolver.LookupHost
+func (trr *TrustedRecursiveResolver2) LookupHost(ctx context.Context, hostname string) ([]string, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
+	addrs, err := reso.LookupHost(ctx, hostname)
+	if err != nil {
+		addrs, err = trr.ResolverSystem.LookupHost(ctx, hostname)
+	}
+	return addrs, err
+}
+
+// LookupHTTPS implements model.Resolver.LookupHTTPS
+func (trr *TrustedRecursiveResolver2) LookupHTTPS(ctx context.Context, domain string) (*model.HTTPSSvc, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
+	out, err := reso.LookupHTTPS(ctx, domain)
+	if err != nil {
+		out, err = trr.ResolverSystem.LookupHTTPS(ctx, domain)
+	}
+	return out, err
+}
+
+// LookupNS implements model.Resolver.LookupNS
+func (trr *TrustedRecursiveResolver2) LookupNS(ctx context.Context, domain string) ([]*net.NS, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
+	out, err := reso.LookupNS(ctx, domain)
+	if err != nil {
+		out, err = trr.ResolverSystem.LookupNS(ctx, domain)
+	}
+	return out, err
+}
+
+// Url returns the configured DoH URL for the TRR2 resolver
```

Comment:
At this point while reading the diff, I am failing to understand why we need this method given that `URL` is already a public attribute of the structure and given that, if we use a `model.Resolver`, we cannot access the `Url` method anyway.

Reply:
Sorry, this was actually some experimental code I forgot to remove. Thanks for pointing this out. I have removed it in the new commit.